src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadataFactoryInterface;
+
+/**
+ * Responsible for the resolving of arguments passed to an action.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+final class ArgumentResolver implements ArgumentResolverInterface
+{
+    private $argumentMetadataFactory;
+
+    /**
+     * @var ArgumentValueResolverInterface[]
+     */
+    private $argumentValueResolvers;
+
+    public function __construct(ArgumentMetadataFactoryInterface $argumentMetadataFactory = null, array $argumentValueResolvers = array())
+    {
+        $this->argumentMetadataFactory = $argumentMetadataFactory;
+        $this->argumentValueResolvers = $argumentValueResolvers;
+    }
+
+    public function getArguments(Request $request, $controller)
+    {
+        $arguments = array();
+
+        foreach ($this->argumentMetadataFactory->createArgumentMetadata($controller) as $metadata) {
+            foreach ($this->argumentValueResolvers as $resolver) {
+                if (!$resolver->supports($request, $metadata)) {
+                    continue;
+                }
+
+                $resolved = $resolver->resolve($request, $metadata);
+
+                if (!$resolved instanceof \Generator) {
```

Comment:
@iltar If we want to make that change, it must be done before the 3.1 release. Otherwise, code relying on `Generator` instances being returned might break when a `Traversable` is returned instead.